File: redtra_customisation/redtra_customisation/page/cheque_lookup/cheque_lookup.js
// Copyright (c) 2026, Redtra Customisation contributors

frappe.pages['cheque-lookup'].on_page_load = (wrapper) => new ChequeLookup(wrapper);

class ChequeLookup {
	constructor(wrapper) {
		this.wrapper = $(wrapper);
		frappe.ui.make_app_page({ parent: wrapper, title: __('Cheque Lookup'), single_column: true });
		this.body = this.wrapper.find('.layout-main-section');
		this.render();
	}

	render() {
		this.body.html(`
		<style>
		.cheque-lookup{max-width:1180px;margin:auto;padding:10px 0 28px}.cheque-hero{padding:28px;border-radius:18px;background:linear-gradient(125deg,#1d3557,#087f75);color:#fff;box-shadow:0 14px 32px #0f172a30}.cheque-hero h2{margin:0 0 5px;color:inherit;font-weight:700}.cheque-hero p{margin:0 0 20px;color:#ffffffcf}.cheque-search{display:flex;gap:10px;max-width:700px}.cheque-search input{flex:1;border:0;border-radius:10px;padding:10px 14px;color:#111}.cheque-search .btn{border-radius:10px;font-weight:600}.cheque-results{margin-top:22px}.cheque-empty{border:1px dashed var(--border-color);border-radius:14px;padding:42px;text-align:center;color:var(--text-muted)}.cheque-card{margin-bottom:20px;border:1px solid var(--border-color);border-radius:16px;overflow:hidden;background:var(--card-bg);box-shadow:0 5px 20px #0f172a10}.cheque-head{display:flex;justify-content:space-between;gap:16px;padding:20px 22px;border-bottom:1px solid var(--border-color)}.cheque-head h3{margin:0 0 4px;font-size:18px}.cheque-amount{text-align:right;font-weight:700;font-size:22px}.cheque-grid{display:grid;grid-template-columns:repeat(4,minmax(0,1fr));gap:12px;padding:20px 22px}.cheque-kv{padding:10px 12px;border-radius:10px;background:var(--fg-hover-color)}.cheque-kv label{display:block;margin-bottom:3px;color:var(--text-muted);font-size:11px;font-weight:600;text-transform:uppercase}.cheque-kv span{font-size:13px;overflow-wrap:anywhere}.cheque-section{padding:0 22px 22px}.cheque-section h4{margin:4px 0 12px;font-size:14px}.cheque-table{border:1px solid var(--border-color);border-radius:10px;overflow:hidden}.cheque-table th{background:var(--fg-hover-color);font-size:11px;text-transform:uppercase;color:var(--text-muted)}.cheque-table td,.cheque-table th{padding:11px 12px}.cheque-payment{display:flex;justify-content:space-between;gap:12px;padding:14px;border:1px solid var(--border-color);border-radius:10px;background:var(--fg-hover-color)}.cheque-action{display:inline-block;margin-top:5px;font-size:12px;font-weight:600;color:var(--primary)}.cheque-expense{font-size:12px;line-height:1.5}.cheque-pill{display:inline-block;padding:3px 9px;border-radius:999px;background:#e2e8f0;color:#334155;font-size:12px;font-weight:600}.cheque-pill.pending{background:#fef3c7;color:#92400e}.cheque-pill.converted,.cheque-pill.paid,.cheque-pill.collected{background:#dcfce7;color:#166534}.cheque-pill.cancelled,.cheque-pill.bounced{background:#fee2e2;color:#b91c1c}@media(max-width:768px){.cheque-search,.cheque-payment,.cheque-head{flex-direction:column}.cheque-grid{grid-template-columns:repeat(2,minmax(0,1fr))}.cheque-amount{text-align:left}}
		</style>
		<div class="cheque-lookup"><div class="cheque-hero"><h2>${__('Find a cheque')}</h2><p>${__('See the post-dated cheque, payment entry and all linked invoices in one place.')}</p><form class="cheque-search"><input placeholder="${__('Enter cheque or reference number')}" autocomplete="off"><button class="btn btn-primary" type="submit">${__('Search')}</button></form></div><div class="cheque-results"><div class="cheque-empty">${__('Enter a cheque number to begin.')}</div></div></div>`);
		this.body.find('.cheque-search').on('submit', (event) => { event.preventDefault(); this.search(); });
	}

	search() {
		const cheque_no = (this.body.find('.cheque-search input').val() || '').trim();
		if (!cheque_no) return frappe.show_alert({ message: __('Enter a cheque number'), indicator: 'orange' });
		const $button = this.body.find('.cheque-search button').prop('disabled', true).text(__('Searching…'));
		this.body.find('.cheque-results').html(`<div class="cheque-empty">${__('Looking up cheque details…')}</div>`);
		frappe.call({ method: 'redtra_customisation.redtra_customisation.page.cheque_lookup.cheque_lookup.get_cheque_details', args: { cheque_no }, callback: (r) => this.results(r.message || {}), always: () => $button.prop('disabled', false).text(__('Search')) });
	}

	results(data) {
		const matches = data.matches || [];
		if (!matches.length) return this.body.find('.cheque-results').html(`<div class="cheque-empty"><strong>${__('No cheque found')}</strong><br>${__('Try the exact cheque/reference number or part of it.')}</div>`);
		this.body.find('.cheque-results').html(`<h3 style="margin:0 0 12px">${matches.length === 1 ? __('Cheque details') : __('{0} cheque records found', [matches.length])}</h3>${matches.map((row) => this.card(row)).join('')}`);
		this.body.find('[data-route]').on('click', (e) => { e.preventDefault(); frappe.set_route(...$(e.currentTarget).data('route').split('|')); });
	}

	card(pdc) {
		const currency = pdc.currency || frappe.defaults.get_default('currency');
		const invoices = (pdc.invoices || []).map((i) => `<tr><td>${this.link(i.doctype,i.name,i.name,i.has_access)}<br>${this.action(i.doctype,i.name,__('Open invoice'),i.has_access)}</td><td>${this.e(i.doctype)}</td><td>${this.project(i.project)}</td><td>${this.e(i.invoice_reference_no || '—')}</td><td>${this.date(i.posting_date)}</td><td class="text-right">${this.money(i.grand_total ?? i.total_amount,i.currency || currency)}</td><td class="text-right">${this.money(i.outstanding_amount ?? i.pdc_outstanding_amount,i.currency || currency)}</td><td>${this.expense(i,currency)}</td><td class="text-right"><strong>${this.money(i.allocated_amount,i.currency || currency)}</strong></td><td>${this.pill(i.status || 'Linked')}</td></tr>`).join('');
		return `<div class="cheque-card"><div class="cheque-head"><div><h3>${this.link('Post Dated Cheques',pdc.name,pdc.reference_no || pdc.name,true)}</h3><div class="text-muted">${this.e(pdc.party_name || pdc.party || '—')} · ${this.e(pdc.payment_type || '—')}</div></div><div class="cheque-amount">${this.money(pdc.amount,currency)}<br>${this.pill(pdc.status || 'Pending')}</div></div><div class="cheque-grid">${this.kv(__('Cheque date'),this.date(pdc.reference_date))}${this.kv(__('Posting date'),this.date(pdc.posting_date))}${this.kv(__('Company'),pdc.company || '—')}${this.kvHtml(__('Project'),this.project(pdc.project))}${this.kv(__('Payment mode'),pdc.mode_of_payment || '—')}${this.kv(__('Bank account'),pdc.bank_account || '—')}${this.kv(__('Party type'),pdc.party_type || '—')}${this.kv(__('PDC reference'),pdc.name)}</div><div class="cheque-section">${this.payment(pdc.payment_entry,currency)}</div><div class="cheque-section"><h4>${__('Linked invoices')} (${(pdc.invoices || []).length})</h4>${invoices ? `<div class="table-responsive"><table class="table cheque-table"><thead><tr><th>${__('Invoice')}</th><th>${__('Type')}</th><th>${__('Project')}</th><th>${__('External Ref.')}</th><th>${__('Date')}</th><th class="text-right">${__('Total')}</th><th class="text-right">${__('Outstanding')}</th><th>${__('Expense')}</th><th class="text-right">${__('Allocated')}</th><th>${__('Status')}</th></tr></thead><tbody>${invoices}</tbody></table></div>` : `<div class="text-muted">${__('No invoices are linked to this cheque.')}</div>`}</div>${pdc.notes ? `<div class="cheque-section"><h4>${__('Notes')}</h4><div class="text-muted">${this.e(pdc.notes)}</div></div>` : ''}</div>`;
	}

	payment(p,currency) { if (!p) return `<h4>${__('Payment entry')}</h4><div class="text-muted">${__('This cheque has not been converted to a payment entry yet.')}</div>`; if (!p.has_access) return `<h4>${__('Payment entry')}</h4><div class="text-muted">${__('A payment entry exists, but you do not have permission to view it.')}</div>`; return `<h4>${__('Payment entry')}</h4><div class="cheque-payment"><div>${this.link('Payment Entry',p.name,p.name,true)}<br><span class="text-muted">${this.e(p.mode_of_payment || '—')} · ${this.date(p.posting_date)}</span></div><div class="text-right">${this.pill(p.status || '—')}<br><strong>${this.money(p.received_amount || p.paid_amount,currency)}</strong></div></div>`; }
	kv(label,value) { return `<div class="cheque-kv"><label>${this.e(label)}</label><span>${this.e(value)}</span></div>`; }
	kvHtml(label,value) { return `<div class="cheque-kv"><label>${this.e(label)}</label><span>${value || '—'}</span></div>`; }
	link(dt,name,label,can) { return can && name ? `<a href="#" data-route="Form|${this.a(dt)}|${this.a(name)}">${this.e(label)}</a>` : this.e(label || '—'); }
	action(dt,name,label,can) { return can && name ? `<a href="#" class="cheque-action" data-route="Form|${this.a(dt)}|${this.a(name)}">${this.e(label)} →</a>` : ''; }
	project(project) { return project ? `${this.link('Project',project,project,true)}<br>${this.action('Project',project,__('Open project'),true)}` : '—'; }
	expense(invoice,currency) { const lines=invoice.expense_lines || []; if (!lines.length) return '—'; return `<div class="cheque-expense">${lines.map((line) => `${this.e(line.expense_account || line.item_name || line.item_code || '—')}<br><strong>${this.money(line.amount,currency)}</strong>`).join('<hr style="margin:5px 0">')}</div>`; }
	pill(value) { const key=String(value || '').toLowerCase().replace(/\s+/g,'-'); return `<span class="cheque-pill ${this.a(key)}">${this.e(value || '—')}</span>`; }
	date(value) { return value ? frappe.datetime.str_to_user(value) : '—'; }
	money(value,currency) { return value === undefined || value === null || value === '' ? '—' : format_currency(value,currency); }
	e(value) { return frappe.utils.escape_html(String(value ?? '')); }
	a(value) { return this.e(value).replace(/&quot;/g,''); }
}

File: redtra_customisation/redtra_customisation/page/cheque_lookup/cheque_lookup.py
"""Single-screen lookup for a cheque's PDC, payment, and linked invoices."""

import frappe
from frappe import _


@frappe.whitelist()
def get_cheque_details(cheque_no: str) -> dict:
	cheque_no = (cheque_no or "").strip()
	if not cheque_no:
		frappe.throw(_("Enter a cheque number."))
	frappe.has_permission("Post Dated Cheques", "read", throw=True)
	rows = frappe.get_list(
		"Post Dated Cheques", filters={"reference_no": ("like", f"%{cheque_no}%")}, fields=["name"],
		order_by="reference_no asc, reference_date desc, modified desc", limit_page_length=25,
	)
	matches = [_pdc_details(row.name) for row in rows]
	# A cheque may have been entered directly as a Payment Entry or as a
	# Security Instrument, without a Post Dated Cheques document.
	if not matches:
		matches = _payment_entry_matches(cheque_no) + _security_instrument_matches(cheque_no)
	return {"query": cheque_no, "matches": matches}


def _payment_entry_matches(cheque_no: str) -> list[dict]:
	if not frappe.has_permission("Payment Entry", "read"):
		return []
	rows = frappe.get_list(
		"Payment Entry",
		filters={"reference_no": ("like", f"%{cheque_no}%"), "docstatus": ("!=", 2)},
		fields=["name"], order_by="posting_date desc, modified desc", limit_page_length=25,
	)
	return [_payment_entry_details(row.name) for row in rows]


def _security_instrument_matches(cheque_no: str) -> list[dict]:
	if not frappe.db.exists("DocType", "Security Instrument") or not frappe.has_permission("Security Instrument", "read"):
		return []
	rows = frappe.get_list(
		"Security Instrument",
		filters={"reference_no": ("like", f"%{cheque_no}%"), "docstatus": ("!=", 2)},
		fields=["name"], order_by="reference_date desc, modified desc", limit_page_length=25,
	)
	return [_security_instrument_details(row.name) for row in rows]


def _payment_entry_details(name: str) -> dict:
	payment = _payment(name) or {}
	invoices = payment.get("invoices", [])
	projects = _unique_projects([payment.get("project")] + [project for invoice in invoices for project in invoice.get("projects", [])])
	return {
		"name": name, "source_doctype": "Payment Entry", "source_label": _("Payment Entry"),
		"status": payment.get("status"), "company": payment.get("company"),
		"project": payment.get("project") or (projects[0] if len(projects) == 1 else None), "projects": projects,
		"party_type": payment.get("party_type"), "party": payment.get("party"), "party_name": payment.get("party"),
		"payment_type": payment.get("payment_type"), "mode_of_payment": payment.get("mode_of_payment"),
		"reference_no": payment.get("reference_no"), "reference_date": payment.get("reference_date"),
		"posting_date": payment.get("posting_date"), "amount": payment.get("received_amount") or payment.get("paid_amount"),
		"currency": payment.get("currency"), "bank_account": payment.get("paid_to") or payment.get("paid_from"),
		"notes": payment.get("remarks"), "payment_entry": payment, "invoices": invoices,
	}


def _security_instrument_details(name: str) -> dict:
	instrument = frappe.get_doc("Security Instrument", name)
	payment = _payment(instrument.payment_entry)
	invoices = (payment or {}).get("invoices", [])
	return {
		"name": name, "source_doctype": "Security Instrument", "source_label": _("Security Instrument"),
		"status": instrument.status, "company": instrument.company, "project": instrument.project,
		"projects": _unique_projects([instrument.project]), "party_type": instrument.party_type,
		"party": instrument.party, "party_name": instrument.party, "payment_type": instrument.payment_type,
		"mode_of_payment": instrument.mode_of_payment, "reference_no": instrument.reference_no,
		"reference_date": instrument.reference_date, "posting_date": instrument.posting_date,
		"amount": instrument.amount, "currency": None, "bank_account": instrument.bank_account,
		"notes": instrument.remarks, "payment_entry": payment, "invoices": invoices,
	}


def _pdc_details(name: str) -> dict:
	pdc = frappe.get_doc("Post Dated Cheques", name)
	payment = _payment(pdc.payment_entry)
	pdc_invoices = [_invoice(row) for row in pdc.get("invoice_references") or []]
	invoices = _merge_invoices(pdc_invoices, (payment or {}).pop("invoices", []))
	projects = _unique_projects(
		[pdc.project] + [project for invoice in invoices for project in invoice.get("projects", [])]
	)
	return {
		"name": pdc.name, "status": pdc.status, "company": pdc.company,
		"project": pdc.project or (projects[0] if len(projects) == 1 else None), "projects": projects,
		"party_type": pdc.party_type, "party": pdc.party, "party_name": pdc.party_name,
		"payment_type": pdc.payment_type, "mode_of_payment": pdc.mode_of_payment,
		"reference_no": pdc.reference_no, "reference_date": pdc.reference_date,
		"posting_date": pdc.posting_date, "amount": pdc.amount, "currency": pdc.account_currency,
		"bank_account": pdc.bank_account, "notes": pdc.notes,
		"payment_entry": payment,
		"invoices": invoices,
	}


def _payment(name: str | None) -> dict | None:
	if not name or not frappe.db.exists("Payment Entry", name):
		return None
	if not frappe.has_permission("Payment Entry", "read", doc=name):
		return {"name": name, "has_access": False}
	fields = [
		"name",
		"company",
		"status",
		"posting_date",
		"reference_no",
		"reference_date",
		"mode_of_payment",
		"party_type",
		"party",
		"paid_from",
		"paid_to",
		"received_amount",
		"paid_amount",
		"paid_to_account_currency",
		"paid_from_account_currency",
		"remarks",
	]
	if frappe.get_meta("Payment Entry").has_field("project"):
		fields.append("project")
	row = frappe.db.get_value("Payment Entry", name, fields, as_dict=True)
	if row:
		row.has_access = True
		row["currency"] = row.get("paid_to_account_currency") or row.get("paid_from_account_currency")
		row["invoices"] = [
			_invoice(reference)
			for reference in frappe.get_all(
				"Payment Entry Reference",
				filters={"parent": name, "parenttype": "Payment Entry"},
				fields=[
					"reference_doctype",
					"reference_name",
					"bill_no",
					"custom_supplier_invoice_no",
					"allocated_amount",
					"total_amount",
					"outstanding_amount",
				],
				order_by="idx asc",
			)
		]
	return row


def _invoice(reference) -> dict:
	doctype, name = reference.get("reference_doctype"), reference.get("reference_name")
	row = {
		"doctype": doctype, "name": name,
		"invoice_reference_no": reference.get("invoice_reference_no") or reference.get("bill_no")
		or reference.get("custom_supplier_invoice_no"),
		"allocated_amount": reference.get("allocated_amount"), "total_amount": reference.get("total_amount"),
		"pdc_outstanding_amount": reference.get("outstanding_amount"), "has_access": False,
	}
	if doctype not in ("Sales Invoice", "Purchase Invoice") or not name:
		return row
	if not frappe.has_permission(doctype, "read", doc=name):
		return row
	invoice = frappe.db.get_value(
		doctype, name,
		["posting_date", "due_date", "grand_total", "outstanding_amount", "status", "project", "currency"],
		as_dict=True,
	)
	if invoice:
		row.update(invoice)
		row["has_access"] = True
		if doctype == "Purchase Invoice":
			row["expense_lines"] = frappe.get_all(
				"Purchase Invoice Item",
				filters={"parent": name, "parenttype": "Purchase Invoice"},
				fields=["item_code", "item_name", "expense_account", "amount", "project"],
				order_by="idx asc",
			)
			row["expense_total"] = sum(line.amount or 0 for line in row["expense_lines"])
		else:
			row["expense_lines"] = []

		# Purchase invoices commonly carry the project at item level rather than
		# on the invoice header. Surface that project in both the invoice row and
		# the cheque summary.
		row["projects"] = _unique_projects(
			[row.get("project")] + [line.get("project") for line in row["expense_lines"]]
		)
		if not row.get("project") and len(row["projects"]) == 1:
			row["project"] = row["projects"][0]
	return row


def _unique_projects(projects: list[str | None]) -> list[str]:
	"""Keep project links ordered and unique, ignoring blank values."""
	return list(dict.fromkeys(project for project in projects if project))


def _merge_invoices(*invoice_groups: list[dict]) -> list[dict]:
	"""Combine PDC and Payment Entry references, keeping one row per invoice."""
	merged = {}
	for invoices in invoice_groups:
		for invoice in invoices:
			key = (invoice.get("doctype"), invoice.get("name"))
			if not all(key):
				continue
			if key in merged:
				# Payment Entry references may have the latest allocation while PDC
				# references contain the external invoice reference.
				for field, value in invoice.items():
					if value not in (None, "", [], 0):
						merged[key][field] = value
			else:
				merged[key] = invoice
	return list(merged.values())
